aptos-move/framework/aptos-framework/sources/aggregator_v2/aggregator_v2.spec.move

```diff
@@ -1,73 +1,106 @@
         pragma opaque;
+        include AbortsIfIntElement<IntElement>;
+        ensures [abstract] result.value == spec_get_value(aggregator);
     }
 
-    spec create_snapshot {
-        // TODO: temporary mockup.
+    spec create_snapshot<IntElement: copy + drop>(value: IntElement): AggregatorSnapshot<IntElement> {
         pragma opaque;
+        include AbortsIfIntElement<IntElement>;
+        ensures [abstract] result.value == value;
     }
 
-    spec read_snapshot {
-        // TODO: temporary mockup.
+    spec read_snapshot<IntElement>(snapshot: &AggregatorSnapshot<IntElement>): IntElement {
         pragma opaque;
+        include AbortsIfIntElement<IntElement>;
+        ensures [abstract] result == snapshot.value;
     }
 
-    spec read_derived_string {
-        // TODO: temporary mockup.
+    spec read_derived_string(snapshot: &DerivedStringSnapshot): String {
         pragma opaque;
+        aborts_if [abstract] false;
+        ensures [abstract] result == snapshot.value;
     }
 
-    spec create_derived_string {
-        // TODO: temporary mockup.
+    spec create_derived_string(value: String): DerivedStringSnapshot {
         pragma opaque;
+        aborts_if [abstract] len(value.bytes) > 1024;
+        ensures [abstract] result.value == value;
     }
 
-    spec derive_string_concat {
-        // TODO: temporary mockup.
+    spec derive_string_concat<IntElement>(before: String, snapshot: &AggregatorSnapshot<IntElement>, after: String): DerivedStringSnapshot {
         pragma opaque;
+        include AbortsIfIntElement<IntElement>;
+        ensures [abstract] result.value.bytes == concat(before.bytes, concat(spec_get_string_value(snapshot).bytes, after.bytes));
+        aborts_if [abstract] len(before.bytes) + len(after.bytes) > 1024;
+    }
+
+    spec schema AbortsIfIntElement<IntElement> {
+        use aptos_std::type_info;
+        aborts_if [abstract] type_info::type_name<IntElement>().bytes != b"u64" && type_info::type_name<IntElement>().bytes != b"u128";
     }
 
     // deprecated
     spec copy_snapshot {
-        // TODO: temporary mockup.
         pragma opaque;
+        aborts_if [abstract] true;
     }
 
     // deprecated
     spec string_concat {
-        // TODO: temporary mockup.
         pragma opaque;
+        aborts_if [abstract] true;
     }
+
+    // Get aggregator.value
+    spec native fun spec_get_value<IntElement>(aggregator: Aggregator<IntElement>): IntElement;
+    // Get aggregator.max_value
+    spec native fun spec_get_max_value<IntElement>(aggregator: Aggregator<IntElement>): IntElement;
+    // Uninterpreted spec function that translates the value inside aggregator into corresponding string representation
+    spec fun spec_get_string_value<IntElement>(aggregator: AggregatorSnapshot<IntElement>): String;
+    spec fun spec_read_snapshot<IntElement>(snapshot: AggregatorSnapshot<IntElement>): IntElement {
```

Comment:
nit: for consistency can also use `get` instead of `read` in these names? So we get aggregator value, get snapshot integer value, etc.

Reply:
I used `read` here because there are corresponding APIs `read_snapshot` and `read_derived_string`.